JIT: Enabled embedded broadcast for binary ops

Ruihan-Yin opened the pull request:
This PR provides the embedded broadcast support for binary ops, to be more specific, ops using the genHWIntrinsic_R_R_RM path.

Including the following ops:
and, andn, or, xor,
min, max,
div, mul, mull, sub,
variable shiftleftlogical/rightarithmetic/rightlogical

Ruihan-Yin committed:
Enabled embedded broadcast for the following ops:
and, andn, or, xor,
min, max,
div, mul, mull, sub,
variable shiftleftlogical/rightarithmetic/rightlogical

Ruihan-Yin committed:
Bug fix:
 JIT used to use a uniform intrinsic for bitwise operations with all data
 types, embedded broadcast is sensitive to input size in this case,
 adding a helper to let emitter aware when input size is long/ulong.

Ruihan-Yin committed:
reset the instruction in the later phase
when embedded broadcast is actually enabled

Ruihan-Yin committed:
filter irrelevant data type in embedded broadcast
There are cases when broadcast node are falsely contained by a embedded
broadcast compatible node, while the data type is actually not supported
Adding extra logics to avoid this situation.

Ruihan-Yin committed:
update the condition on instruction reset:
instructions with either long or ulong as basetype should be reset to
qword instructions.

Ruihan-Yin commented:
Hi @tannergooding , would you please take a look at this PR, 
it mostly includes:
1. some updates on the metadata in the related tables to enable embedded broadcast in more instructions.
2.  some workaround on the bitwise instructions to make it work correctly with embedded broadcast when the input data type is in 64-bit. 
3. a bug fix where some broadcast nodes with data type less than 32 bits are accidentally contained, made some changes in the contain check to filter out these situations.

Ruihan-Yin committed:
Apply format patch

tannergooding commented on src/coreclr/jit/instr.cpp:
```diff
@@ -1230,7 +1230,8 @@ void CodeGen::inst_RV_RV_TT(
     {
         instOptions = INS_OPTS_EVEX_b;
-        if (HWIntrinsicInfo::IsBitwiseOperation(ins) && (op2->AsHWIntrinsic()->Op(1)->TypeIs(TYP_LONG) || op2->AsHWIntrinsic()->Op(1)->TypeIs(TYP_ULONG)))
+        if (HWIntrinsicInfo::IsBitwiseOperation(ins) &&
+            (op2->AsHWIntrinsic()->Op(1)->TypeIs(TYP_LONG) || op2->AsHWIntrinsic()->Op(1)->TypeIs(TYP_ULONG)))
```
You can check both simultaneously with:
```suggestion
        if (HWIntrinsicInfo::IsBitwiseOperation(ins) && varTypeIsLong(op2->AsHWIntrinsic()->Op(1)))
```

tannergooding replied:
That being said, would this be better:
```suggestion
        if (HWIntrinsicInfo::IsBitwiseOperation(ins) && varTypeIsLong(op2->AsHWIntrinsic()->GetSimdBaseType()))
```

Since it checks the exact type of broadcast intended, irrespective of the type of the underlying operand.

Ruihan-Yin replied:
Indeed, I missed the basetype information on broadcast node, thanks for the suggestion, updated.

Ruihan-Yin committed:
Resolve reviews:
make the typecheck based on broadcast node it self.

Ruihan-Yin commented:
## windows x64
Diffs are based on <span style="color:#1460aa">1,627,004</span> contexts (<span style="color:#1460aa">467,427</span> MinOpts, <span style="color:#1460aa">1,159,577</span> FullOpts).

<span style="color:#d35400">MISSED</span> contexts: <span style="color:#d35400">2,227 (0.14%)</span>


<details>
<summary>Overall (<span style="color:red">+2,406</span> bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|aspnet.run.windows.x64.checked.mch|44,062,516|<span style="color:red">+156</span>|
|benchmarks.run.windows.x64.checked.mch|7,244,558|<span style="color:red">+152</span>|
|benchmarks.run_pgo.windows.x64.checked.mch|26,963,808|<span style="color:red">+143</span>|
|benchmarks.run_tiered.windows.x64.checked.mch|10,507,436|<span style="color:red">+157</span>|
|coreclr_tests.run.windows.x64.checked.mch|385,009,583|<span style="color:red">+736</span>|
|libraries.pmi.windows.x64.checked.mch|59,506,418|<span style="color:red">+190</span>|
|libraries_tests.pmi.windows.x64.checked.mch|121,311,207|<span style="color:red">+220</span>|
|realworld.run.windows.x64.checked.mch|13,650,175|<span style="color:red">+652</span>|


</div></details>

<details>
<summary>MinOpts (<span style="color:red">+383</span> bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|benchmarks.run_pgo.windows.x64.checked.mch|10,473,053|<span style="color:red">+73</span>|
|benchmarks.run_tiered.windows.x64.checked.mch|7,567,856|<span style="color:red">+73</span>|
|coreclr_tests.run.windows.x64.checked.mch|269,424,110|<span style="color:red">+237</span>|


</div></details>

<details>
<summary>FullOpts (<span style="color:red">+2,023</span> bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|aspnet.run.windows.x64.checked.mch|29,278,684|<span style="color:red">+156</span>|
|benchmarks.run.windows.x64.checked.mch|6,870,267|<span style="color:red">+152</span>|
|benchmarks.run_pgo.windows.x64.checked.mch|16,490,755|<span style="color:red">+70</span>|
|benchmarks.run_tiered.windows.x64.checked.mch|2,939,580|<span style="color:red">+84</span>|
|coreclr_tests.run.windows.x64.checked.mch|115,585,473|<span style="color:red">+499</span>|
|libraries.pmi.windows.x64.checked.mch|57,987,107|<span style="color:red">+190</span>|
|libraries_tests.pmi.windows.x64.checked.mch|114,958,091|<span style="color:red">+220</span>|
|realworld.run.windows.x64.checked.mch|12,500,933|<span style="color:red">+652</span>|


</div></details>

Ruihan-Yin commented:
Fails should be irrelevant.

We had regression of about 2,000 bytes in the code size introduced by the changes. This should be expected as the regression is the conversion from VEX encoding to EVEX encoding when enabling embedded broadcast.
And there is no influence on the throughput.

Do we accept this results?

tannergooding commented:
> Do we accept the results?

The results look good/within reason to me. There are two important bits to call out...

First this change, in the typical case, trades a slight increase in code size (typically +2-bytes) for a decrease in data size (going from `sizeof(VectorXXX<T>)` to `sizeof(T)`) thereby increasing data locality and minimizing impact to the cache. This will often be a performance win.

Second, SPMI diffs are not currently tracking/including the data size allocations as part of "bytes of code" metric. `HardwareIntrinsics.RayTracer.Packet256Tracer:Shade` for example is `+2 bytes of code`, but `-32 bytes of data` so its actually a net win of `-30 bytes of allocations`

tannergooding commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -8138,11 +8138,18 @@ bool Lowering::IsContainableHWIntrinsicOp(GenTreeHWIntrinsic* parentNode, GenTre
         case NI_AVX512F_BroadcastScalarToVector512:
         {
+            var_types baseType = hwintrinsic->GetSimdBaseType();
+            if (baseType == TYP_BYTE || baseType == TYP_UBYTE || baseType == TYP_SHORT || baseType == TYP_USHORT)
```
This is better handled as
```suggestion
            if (varTypeIsSmall(baseType))
```

tannergooding commented:
CC. @dotnet/jit-contrib for secondary review.

Ruihan-Yin committed:
Resolve review:
use `varTypeIsSmall` type check to cover all the unsupported data type
in embedded broadcast.

BruceForstall commented:
superpmi-replay and superpmi-diffs pipelines are failing with messages like:
```
[17:40:21] ERROR: Couldn't load base metrics summary created by child process
[17:40:21] General fatal error
```

I wonder if the JIT is crashing (not asserting) with this PR?

Ruihan-Yin commented:
> I wonder if the JIT is crashing (not asserting) with this PR?

First time seeing this fail in this PR, the recent changes might not cause the crashing, I will rebase the PR and re-run the CI again to confirm.

BruceForstall commented on src/coreclr/jit/instrsxarch.h:
```diff
@@ -399,7 +399,6 @@ INST3(aesenclast,       "aesenclast",       IUM_WR, BAD_CODE,     BAD_CODE,
 INST3(aesimc,           "aesimc",           IUM_WR, BAD_CODE,     BAD_CODE,     SSE38(0xDB),                             INS_TT_NONE,                                          REX_WIG      | Encoding_VEX)                                                                                                                                                   // Perform the AES InvMixColumn Transformation
 INST3(aeskeygenassist,  "aeskeygenassist",  IUM_WR, BAD_CODE,     BAD_CODE,     SSE3A(0xDF),                             INS_TT_NONE,                                          REX_WIG      | Encoding_VEX)                                                                                                                                                   // AES Round Key Generation Assist
 INST3(pclmulqdq,        "pclmulqdq" ,       IUM_WR, BAD_CODE,     BAD_CODE,     SSE3A(0x44),                             INS_TT_NONE,                         Input_64Bit    | REX_WIG      | Encoding_VEX                   | INS_Flags_IsDstDstSrcAVXInstruction)                                                                                           // Perform a carry-less multiplication of two qua ...
-
```
nit: no need to delete this empty line?

tannergooding commented:
> I wonder if the JIT is crashing (not asserting) with this PR?

There are similar failures in multiple PRs from what I've seen recently.

Noting that SPMI diffs is faulting in Python when processing the diffs summary:
```
Traceback (most recent call last):
  File "C:\h\w\B2080989\p\superpmi.py", line 4667, in <module>
    sys.exit(main(args))
  File "C:\h\w\B2080989\p\superpmi.py", line 4558, in main
    success = asm_diffs.replay_with_asm_diffs()
  File "C:\h\w\B2080989\p\superpmi.py", line 2019, in replay_with_asm_diffs
    self.write_asmdiffs_markdown_summary(write_fh, asm_diffs)
  File "C:\h\w\B2080989\p\superpmi.py", line 2170, in write_asmdiffs_markdown_summary
    write_row(*t)
  File "C:\h\w\B2080989\p\superpmi.py", line 2165, in write_row
    num_missed_base / num_contexts * 100,
ZeroDivisionError: division by zero
```

Replay is similarly failing with things like:
```
[17:42:50] Invoking: C:\h\w\B6DA095E\p\superpmi.exe -v ewi -r C:\h\w\B6DA095E\t\tmpdsiibugo\repro -p -jitoption JitStressRegs=0x80 -f C:\h\w\B6DA095E\t\tmpdsiibugo\coreclr_tests.run.windows.x64.checked.mch_fail.mcl -metricsSummary C:\h\w\B6DA095E\t\tmpdsiibugo\coreclr_tests.run.windows.x64.checked.mch_metrics.csv C:\h\w\B6DA095E\p\clrjit_win_x64_x64.dll C:\h\w\B6DA095E\p\artifacts\spmi\mch\02e334af-4e6e-4a68-9feb-308d3d2661bc.windows.x64\coreclr_tests.run.windows.x64.checked.mch
[17:42:50] ERROR: Couldn't load base metrics summary created by child process
[17:42:50] ERROR: Couldn't load base metrics summary created by child process
[17:42:50] ERROR: Couldn't load base metrics summary created by child process
[17:42:50] ERROR: Couldn't load base metrics summary created by child process
[17:42:50] General fatal error
[17:42:50] Running SuperPMI replay of C:\h\w\B6DA095E\p\artifacts\spmi\mch\02e334af-4e6e-4a68-9feb-308d3d2661bc.windows.x64\libraries.crossgen2.windows.x64.checked.mch
```

Did we recently get a JIT/EE version change and potentially not have up to date baselines?

BruceForstall commented:
> Noting that SPMI diffs is faulting in Python when processing the diffs summary:

superpmi.py should gracefully handle that. But I think that's a symptom that superpmi.exe crashed.

> Did we recently get a JIT/EE version change and potentially not have up to date baselines?

If the GUID changed, we wouldn't have this problem because we would only pick up matched MCH files.

If someone changed the interface without changing the GUID, then anything is possible.

BruceForstall commented:
I kicked off runs on `main` (not on a PR) so we can see if there are failures in the main branch:

https://dev.azure.com/dnceng-public/public/_build/results?buildId=331428&view=results
https://dev.azure.com/dnceng-public/public/_build/results?buildId=331429&view=results

BruceForstall commented:
@Ruihan-Yin The superpmi pipelines are now running clean. I would suggest to merge up to HEAD and re-push your PR to trigger re-run of these pipelines.

Ruihan-Yin commented:
> The superpmi pipelines are now running clean. I would suggest to merge up to HEAD and re-push your PR to trigger re-run of these pipelines.
Thanks for the notification, will resolve the reviews and rebase the branch soon.

Ruihan-Yin committed:
Resolve reviews:
1. put the IsBitwiseInstruction to a proper place.
2. nit: restored unnecessary line delete.

Ruihan-Yin commented on src/coreclr/jit/hwintrinsic.h:
```diff
@@ -598,6 +598,23 @@ struct HWIntrinsicInfo
     }
 
+#if defined(TARGET_XARCH)
+    static bool IsBitwiseOperation(instruction ins)
```
moved the function into emitxarch.h/cpp as a static function.

Ruihan-Yin commented:
Hi @BruceForstall, some tests are cancelled for some reasons, I suppose it is not relevant to the changes in this PR, shall I re-run the test again? And can you please review the new changes, if any. Thanks!

tannergooding commented on src/coreclr/jit/instr.cpp:
```diff
@@ -1230,6 +1230,30 @@ void CodeGen::inst_RV_RV_TT(
     if (IsEmbBroadcast)
     {
         instOptions = INS_OPTS_EVEX_b;
+        if (emitter::IsBitwiseInstruction(ins) && varTypeIsLong(op2->AsHWIntrinsic()->GetSimdBaseType()))
```
Possibly not. I'll let Ruihan-Yin fix this in a follow up PR if they want to, that way we can land the important bit before the snap.

Ruihan-Yin replied:
Thanks for the feedback, we will fix this point in the following PR on embedded broadcast.
